qubit-sdk/src/main/java/com/qubit/android/sdk/internal/placement/interactor/PlacementInteractorImpl.kt

```diff
@@ -38,24 +41,54 @@ internal class PlacementInteractorImpl(
       onSuccess: OnPlacementSuccess,
       onError: OnPlacementError
   ) {
+    val placementMode = mode ?: DEFAULT_PLACEMENT_MODE
+    val attributes = buildAttributesJson(deviceId)
+    val cacheKey = buildCacheKey(placementId, placementMode, previewOptions, attributes)
     placementConnector.getPlacementModel(
         getPlacementApiHost(configurationRepository),
         placementId,
-        mode ?: DEFAULT_PLACEMENT_MODE,
-        deviceId,
+        placementMode,
         previewOptions,
-        { handleSuccessfulResponse(it, onSuccess, onError) },
-        { onError(it) }
+        attributes,
+        { handleSuccessfulResponse(it, cacheKey, true, onSuccess, onError) },
+        { handleErrorResponse(it, cacheKey, onSuccess, onError) }
     )
   }
 
+  private fun buildAttributesJson(
+      deviceId: String
+  ): JsonObject {
+    return JsonObject().apply {
+      add("visitor", JsonObject().apply {
+        addProperty("id", deviceId)
+      })
+    }
+  }
+
+  private fun buildCacheKey(
+      placementId: String,
+      placementMode: PlacementMode,
+      previewOptions: PlacementPreviewOptions,
+      attributes: JsonObject
+  ): String {
+    return StringBuilder().apply {
+      append(placementId)
```

Comment:
It is good practice to separate elements in composite key.
For example: without separators, the same key will be produced in both cases:
1. compaignId=11, experienceId=1
2. compaignId=1, experienceId=11

I know, that in this particular case it can be unlikely to happen, if ids are UUIDs.